### PR TITLE
Drop futures as soon as they're done for `array::join`

### PR DESCRIPTION
**Pull request by yoshuawuyts**
Makes progress towards #137. If we accept this fix we'll want to copy it for all our APIs.

### CODE REVIEW

File: src/future/join/array.rs [hunk 1]
```diff
@@ -29,7 +29,7 @@ where
     wakers: WakerArray<N>,
     state: PollArray<N>,
     #[pin]
-    futures: [Fut; N],
+    futures: [MaybeUninit<Fut>; N],
```
Comment (jstarks) [inline src/future/join/array.rs:32]: Would `ManuallyDrop` be clearer?

File: src/future/join/array.rs [hunk 2]
```diff
@@ -94,20 +94,32 @@ where
...
+                        .map_unchecked_mut(|t| t.assume_init_mut())
+                        .poll(&mut cx)
+                } {
                     this.items[i] = MaybeUninit::new(value);
                     this.state[i].set_ready();
                     *this.pending -= 1;
                 }
 
+                // If the state was changed from "pending" to "ready", drop the future.
+                if this.state[i].is_ready() {
```
Comment (jstarks) [inline src/future/join/array.rs:118]: Seems like this could merge into the previous `if` block.

File: src/future/join/array.rs (hunk 1 above)
Comment (yoshuawuyts) [reply src/future/join/array.rs:32]: Yes, that's a great suggestion — it would be!

File: src/future/join/array.rs (hunk 2 above)
Comment (yoshuawuyts) [reply src/future/join/array.rs:118]: I initially had that, but it gave a lifetime error. Let me change it to use ManuallyDrop and try again.

**Comment by yoshuawuyts** [comment]
Updated with feedback from review! I'm pretty happy with the logic as it is, and I think we could merge it. But I think I'd be happiest if we could factor out the manual `[ManuallyDrop<T>; N]` logic into its own file, because we'll want to share it between all array implementations.

**Pull request merged**